core/jvm/src/test/scala/zio/IOSpec.scala

```diff
@@ -279,6 +293,57 @@ class IOSpec(implicit ee: org.specs2.concurrent.ExecutionEnv) extends TestRuntim
     }
   }
 
+  def testSomeOrFailWithNone = {
+    val task: Task[Int] = UIO(Option.empty[Int]).someOrFail(exampleError)
+    unsafeRun(task) must throwA[FiberFailure]
+  }
+
+  def testSomeOrFailExtractOptionalValue = {
+    val task: Task[Int] = UIO(Some(42)).someOrFail(exampleError)
+    unsafeRun(task) must_=== 42
+  }
+
+  def testSomeOrFailExceptionOnOptionalValue = unsafeRun(ZIO.succeed(Some(42)).someOrFailException) must_=== 42
+
+  def testSomeOrFailExceptionOnEmptyValue = {
+    val task = ZIO.succeed(Option.empty[Int]).someOrFailException
+    unsafeRun(task) must throwA[FiberFailure]
+  }
+
+  def testRightOrFailExceptionOnRightValue = unsafeRun(ZIO.succeed(Right(42)).rightOrFailException) must_=== 42
+
+  def testRightOrFailExceptionOnLeftValue = {
+    val task: Task[Int] = ZIO.succeed(Left(2)).rightOrFailException
+    unsafeRun(task) must throwA[FiberFailure]
+  }
+
+  def testRightOrFailExtractsRightValue = {
+    val task: Task[Int] = UIO(Right(42)).rightOrFail(exampleError)
+    unsafeRun(task) must_=== 42
+  }
+
+  def testRightOrFailWithLeft = {
+    val task: Task[Int] = UIO(Left(1)).rightOrFail(exampleError)
+    unsafeRun(task) must throwA[FiberFailure]
+  }
+
+  def testLeftOrFailExceptionOnLeftValue = unsafeRun(ZIO.succeed(Left(42)).leftOrFailException) must_=== 42
+
+  def testLeftOrFailExceptionOnRightValue = {
+    val task: Task[Int] = ZIO.succeed(Right(2)).leftOrFailException
```

Comment:
Does this work without annotating `task` as `Task[Int]`?

Reply:
Great question. No.
The return type by the compiler is ```[B, C, E1 >:NoSuchElementException]zio.ZIO[Any,E1,B]```. 🤔 

also, changing the signature to 
```scala
def leftOrFailException[B, C](
    implicit ev: A <:< Either[B, C]
  ): ZIO[R, NoSuchElementException, B]```
does not work too.

Reply:
Actually, I tested and it looks like it infers properly - scalac says it's a `ZIO[Any, NoSuchElementException, Nothing]` which is accurate.